### PR TITLE
feat: Add allowlist of GitHub teams for OAuth

### PR DESCRIPTION
**Pull request by kylecarbs**
Fixes #2848.

### CODE REVIEW

File: cli/server.go [hunk 1]
```diff
@@ -719,7 +722,7 @@ func configureTLS(listener net.Listener, tlsMinVersion, tlsClientAuth, tlsCertFi
 	return tls.NewListener(listener, tlsConfig), nil
 }
 
-func configureGithubOAuth2(accessURL *url.URL, clientID, clientSecret string, allowSignups bool, allowOrgs []string) (*coderd.GithubOAuth2Config, error) {
+func configureGithubOAuth2(accessURL *url.URL, clientID, clientSecret string, allowSignups bool, allowOrgs []string, allowTeams []string) (*coderd.GithubOAuth2Config, error) {
```
Comment (Emyrk) [inline cli/server.go:725]: We check the `allowTeams` is formatted correctly when a user tries to login. So if the deployment has an issue, we don't know until login. Should we verify format up front in this function?

If the format is incorrect (not 2 parts), should we just log an error and kill coder from starting?

Comment (Emyrk) [reply cli/server.go:725]: On that note, should we pass this as a datastruct so we don't have to parse this again? That might be overkill, just would be unfortunate to have to strings.Split again and check for parts.

File: coderd/userauth.go [hunk 2]
```diff
@@ -64,6 +67,49 @@ func (api *API) userOAuth2Github(rw http.ResponseWriter, r *http.Request) {
...
+
+		var allowedTeam *github.Team
+		for _, team := range teams {
+			for _, organizationAndTeam := range api.GithubOAuth2Config.AllowTeams {
+				parts := strings.SplitN(organizationAndTeam, "/", 2)
+				if len(parts) != 2 {
+					httpapi.Write(rw, http.StatusInternalServerError, httpapi.Response{
+						Message: "Team allowlist isn't formatted correctly.",
+						Detail:  fmt.Sprintf("Got %s, wanted <organization>/<team>", organizationAndTeam),
+					})
```
Comment (Emyrk) [inline coderd/userauth.go:89]: I mentioned we should check this earlier, but if we are ok with delaying the parse until here, we should also api.Logger.Error log this failure.

**Review by Emyrk — COMMENTED** [review]
Minor nit about where to parse the allow teams

File: cli/server.go (hunk 1 above)
Comment (kylecarbs) [reply cli/server.go]: I was hesitant about this at first, but I think it turned out cleaner this way. Good suggestion!

**Comment by kylecarbs** [comment]
@Emyrk fixed up!

**Pull request merged**